### DISCUSSION TITLE
Calling a C++ Library in .udf file

### DISCUSSION THREAD

**Discussion by raj-brown** [Nek5000/nekRS#585] [General]
Dear All-
I am trying to call an external C++ library win .udf file but while import .h files It complains that I can not fine it. Although I compiled nekRS by adding include dir and linking the library. Still It could not find .h files. Could you please help me with this.

Thanks!
Raj

**Comment by stgeke** [comment]
The udf file is JIT compiled, so adding the include s+ libs when building nekRS won't work. 
Please define the following env-vars to make it work:

`NEKRS_UDF_INCLUDES` for optional include directories ("/path1/foo1;/path2/foo2")
`NEKRS_UDF_LIBS` for optional libraries ("/path1/lib1.so;/path2/lib2.so")

**Reply by raj-brown** [reply]
hi @stgeke Thank you! I will try these.

**Reply by raj-brown** [reply]
@stgeke One question: The include env-var will work if include directory has a lost of subdirs?

**Reply by raj-brown** [reply]
hi @stgeke I tried defining library and include files via environment variables but it did not work.

**Reply by stgeke** [reply]
Make sure to use latest `next`.

**Reply by raj-brown** [reply]
@stgeke I did not get, what is next.

**Reply by stgeke** [reply]
The `next` branch instead of the default (`master`).

**Comment by raj-brown** [comment]
@stgeke will you have a quick 10 mnts? I can show you the error

**Comment by raj-brown** [comment]
Hi @stgeke I tried next branch and it did not work either.

**Comment by raj-brown** [comment]
Hi @stgeke when i run an example from next branch it gives the following errors
'''
/users/kshukla1/.local/nekrs/include/elliptic/elliptic.h:159:10: fatal error: ellipticMultiGrid.h: No such file or directory
  159 | #include "ellipticMultiGrid.h"
      |          ^~~~~~~~~~~~~~~~~~~~~
'''
Could you please help me with this.

**Reply by stgeke** [reply]
What example are we talking about?

**Comment by raj-brown** [comment]
I was running turbPipe

**Reply by stgeke** [reply]
Running turbPipe fails even if you don't set `NEKRS_UDF_INCLUDES `?

**Comment by raj-brown** [comment]
Yes! without modifying UDF_INCLUDES

**Comment by stgeke** [comment]
Did you remove the build and installation dir before building the latest `next` branch?

**Comment by raj-brown** [comment]
yes..i did from scratch

**Reply by stgeke** [reply]
Does `ellipticMultiGrid.h` exist in the installation directory?

**Comment by raj-brown** [comment]
yes
but it is in MG folder

**Comment by raj-brown** [comment]
Hi @stgeke Do I need to change something in CMAke?

**Comment by raj-brown** [comment]
Hi @stgeke I was able to build the ```next``` branch and ran an existing example successfully...but defining UDF_INCLUDE did not work.Any suggestion would be really helpful..Thanks!

**Reply by stgeke** [reply]
Can you please post how you defined `NEKRS_UDF_INCLUDES`.

**Comment by raj-brown** [comment]
Hi Stefan @stgeke if you have time for a quick zoom then that will be really helpful. I also wanted to discuss about this paper of yours https://arxiv.org/pdf/2409.06404. Thank you very much!

**Reply by stgeke** [reply]
Please DM me.

**Reply by raj-brown** [reply]
Thanks Stefan @stgeke.Could you please provide me your email id. Thank you!

**Comment by stgeke** [comment]
You'll find it here: https://nek5000.mcs.anl.gov/

**Reply by raj-brown** [reply]
Hi @stgeke I just ent you email..

**Comment by raj-brown** [comment]
Hi @stgeke this is how I define the INCLUDE neviron-
```
export NEKRS_UDF_INCLUDES=$NEKRS_UDF_INCLUDES:/oscar/data/gk/kshukla1/CFD_NEKRS/nekRS/3rd_party/pytorch/libtorch/include
```

**Reply by stgeke** [reply]
Please try
```
export NEKRS_UDF_INCLUDES=/oscar/data/gk/kshukla1/CFD_NEKRS/nekRS/3rd_party/pytorch/libtorch/include
```

**Reply by stgeke** [reply]
Please also remove the udf cache
```
rm -rf .cache/udf
```

**Reply by raj-brown** [reply]
Hi @stgeke I just did and it did not work
```
In file included from /oscar/data/gk/kshukla1/CFD_NEKRS/nekRS/examples/turbPipe/.cache/udf/udf.cpp:7:
/oscar/data/gk/kshukla1/CFD_NEKRS/nekRS/examples/turbPipe/turbPipe.udf:4:10: fatal error: torch/torch.h: No such file or directory
    4 | #include "torch/torch.h"
```

**Comment by raj-brown** [comment]
Hi @stgeke  This is the library which I want to call in ```.udf``` file..
```
https://download.pytorch.org/libtorch/nightly/cpu/libtorch-cxx11-abi-static-without-deps-1.9.0.dev20210516%2Bcpu.zip
```

**Comment by stgeke** [comment]
Hmm it works for me. Not sure what’s going on. Anyway you can also try to
add a udf.cmake to your case directory defining custom includes and libs.

**Reply by raj-brown** [reply]
Hi @stgeke you mean the torch library works for you.Could you please share your case file for .udf.

**Reply by stgeke** [reply]
No, I am talking about the general functionality.

**Comment by raj-brown** [comment]
hi @stgeke this is how i was able to call the library in other C++ library
```
FIND_PACKAGE(Torch REQUIRED)
SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${TORCH_CXX_FLAGS}")
TARGET_LINK_LIBRARIES(SolverUtils LINK_PUBLIC "${TORCH_LIBRARIES}")
```

**Reply by stgeke** [reply]
Sure, you can add something like this to `udf.cmake`

```
FIND_PACKAGE(Torch REQUIRED)
target_link_libraries(udf PUBLIC "${TORCH_LIBRARIES}")
```

**Comment by raj-brown** [comment]
Hi @stgeke It seems like it is working now as I included the .cmake files. I will have to test a couple of more things but I think Its okay now. Thank you Stefan! Also about nekCRF I have sent you an email..Thank you very much!

**Reply by raj-brown** [reply]
Hi @stgeke I was wondering if there is any example where boundary (essential and natural) conditions are varying with time? Thank you very much for your help.

**Comment by raj-brown** [comment]
Hi @stgeke Just wanted to check if you have any example where time dependent boundary condition is implemented. Thank you!

**Reply by stgeke** [reply]
Although I don't recall any examples that use a time-varying boundary condition, it is fully supported. You can simply access the time via `bc->time` within the boundary function of choice.

**Comment by raj-brown** [comment]
hi @stgeke I had a question. I want to solve this adjoint NS equation in NekRS new branch. Below is the equation. I need to pass $\bar{u}$ as input field and I will appreciate if you have any suggestion on how to achieve this. Thank you very much!
![image](https://github.com/user-attachments/assets/c7912944-9750-4da2-8100-c1e02a946653)

**Reply by stgeke** [reply]
Is your question about the integration of adjoints still related to this thread (calling a C++ lib)?

**Reply by raj-brown** [reply]
Hi @stgeke No.It is a different question. Should I open a new thread.?

Thanks!
